multimap/src/main/java/org/infinispan/multimap/impl/HashMapBucket.java

```diff
@@ -116,22 +123,30 @@ public boolean containsKey(K key) {
    /**
     * We do not use the original replace method here. Our implementation allows to create and delete entries.
     */
-   public boolean replace(K key, V expected, V replacement) {
+   public HashMapBucket<K, V> replace(K key, V expected, V replacement) {
       MultimapObjectWrapper<K> storeKey = new MultimapObjectWrapper<>(key);
       V current = values.get(storeKey);
-      if (current == null && expected == null && replacement != null) {
-         values.put(storeKey, replacement);
-         return true;
-      }
 
-      if (!equalValues(current, expected)) return false;
+      if (!equalValues(current, expected)) return null;
+      if (equalValues(current, replacement)) return this;
 
-      if (replacement == null) {
-         values.remove(storeKey);
-      } else {
-         values.put(storeKey, replacement);
+      if (values.isEmpty()) {
+         Map<MultimapObjectWrapper<K>, V> copied = new HashMap<>();
+         copied.put(storeKey, replacement);
+         return new HashMapBucket<>(copied);
       }
-      return true;
+
+      Map<MultimapObjectWrapper<K>, V> copied = new HashMap<>(values.size());
+      for (Map.Entry<MultimapObjectWrapper<K>, V> entry : values.entrySet()) {
```

Comment:
Thinking about HashMap in particular I don't think it is worth it to iterate and do equality check on every key. I think just doing a copy of the existing map and then doing the replace/remove would be faster.